overlord/snapstate/snapstate_test.go

```diff
@@ -1807,6 +1841,268 @@ func (s *snapmgrTestSuite) TestUpdateRunThrough(c *C) {
+				name:     snapName,
+			}, Commentf(snapName))
+			di++
+		}
+	}
+	// we check all snaps with each user
+	c.Check(seen, DeepEquals, map[snapIDuserID]bool{
+		snapIDuserID{snapID: "core-snap-id", userID: 1}:     true,
+		snapIDuserID{snapID: "some-snap-id", userID: 1}:     true,
+		snapIDuserID{snapID: "services-snap-id", userID: 1}: true,
+		snapIDuserID{snapID: "core-snap-id", userID: 2}:     true,
+		snapIDuserID{snapID: "some-snap-id", userID: 2}:     true,
+		snapIDuserID{snapID: "services-snap-id", userID: 2}: true,
+	})
+}
+
+func (s *snapmgrTestSuite) TestUpdateManyMultipleCredsUserRunThrough(c *C) {
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	snapstate.Set(s.state, "core", &snapstate.SnapState{
+		Active: true,
+		Sequence: []*snap.SideInfo{
+			{RealName: "core", Revision: snap.R(1), SnapID: "core-snap-id"},
+		},
+		Current:  snap.R(1),
+		SnapType: "os",
+	})
+	snapstate.Set(s.state, "some-snap", &snapstate.SnapState{
+		Active: true,
+		Sequence: []*snap.SideInfo{
+			{RealName: "some-snap", Revision: snap.R(5), SnapID: "some-snap-id"},
+		},
+		Current:  snap.R(5),
+		SnapType: "app",
+		UserID:   1,
+	})
+	snapstate.Set(s.state, "services-snap", &snapstate.SnapState{
+		Active: true,
+		Sequence: []*snap.SideInfo{
+			{RealName: "services-snap", Revision: snap.R(2), SnapID: "services-snap-id"},
+		},
+		Current:  snap.R(2),
+		SnapType: "app",
+		UserID:   2,
+	})
+
+	chg := s.state.NewChange("refresh", "refresh all snaps")
+	updated, tts, err := snapstate.UpdateMany(s.state, nil, 2)
+	c.Assert(err, IsNil)
+	for _, ts := range tts {
+		chg.AddAll(ts)
+	}
+	c.Check(updated, HasLen, 3)
+
+	s.state.Unlock()
+	defer s.snapmgr.Stop()
+	s.settle(c)
+	s.state.Lock()
+
+	c.Assert(chg.Status(), Equals, state.DoneStatus)
+	c.Assert(chg.Err(), IsNil)
+
+	macaroonMap := map[string]string{
+		"core":          "macaroon2",
+		"some-snap":     "macaroon",
+		"services-snap": "macaroon2",
+	}
+
+	type snapIDuserID struct {
+		snapID string
+		userID int
+	}
+	seen := make(map[snapIDuserID]bool)
+	di := 0
+	for _, op := range s.fakeBackend.ops {
+		switch op.op {
+		case "storesvc-list-refresh":
+			snapID := op.cand.SnapID
+			seen[snapIDuserID{snapID: snapID, userID: op.userID}] = true
+		case "storesvc-download":
+			snapName := op.name
+			c.Check(s.fakeStore.downloads[di], DeepEquals, fakeDownload{
+				macaroon: macaroonMap[snapName],
+				name:     snapName,
+			}, Commentf(snapName))
+			di++
+		}
+	}
+	// we check all snaps with each user
+	c.Check(seen, DeepEquals, map[snapIDuserID]bool{
+		snapIDuserID{snapID: "core-snap-id", userID: 1}:     true,
+		snapIDuserID{snapID: "some-snap-id", userID: 1}:     true,
+		snapIDuserID{snapID: "services-snap-id", userID: 1}: true,
+		snapIDuserID{snapID: "core-snap-id", userID: 2}:     true,
+		snapIDuserID{snapID: "some-snap-id", userID: 2}:     true,
+		snapIDuserID{snapID: "services-snap-id", userID: 2}: true,
+	})
+
+	var coreState, snapState snapstate.SnapState
```

Comment:
Curious, why is this testing "core" and "some-snap" state but not "service-snap" state?

Reply:
because some-snap and service-snap behave the same (keep the user) but I can add it

Reply:
though for services-snap that is not visible because user is already 2 there (the user we pass to UpdateMany)

Reply:
I added comment instead